pkg/util/term/term.go

```diff
@@ -90,7 +97,88 @@ func (t TTY) Safe(fn SafeFunc) error {
 	if err != nil {
 		return err
 	}
-	return interrupt.Chain(t.Parent, func() { term.RestoreTerminal(inFd, state) }).Run(fn)
+	return interrupt.Chain(t.Parent, func() {
+		close(t.stopResizing)
+		term.RestoreTerminal(inFd, state)
+	}).Run(fn)
+}
+
+type Size struct {
+	Width  uint16
+	Height uint16
+}
+
+func (t TTY) GetSize() *Size {
+	if t.In == nil {
+		return nil
+	}
+
+	in := t.In
+
+	desc, ok := in.(fd)
+	if !ok {
+		return nil
+	}
+
+	winsize, err := term.GetWinsize(desc.Fd())
+	if err != nil {
+		return nil
+	}
+
+	return &Size{Width: winsize.Width, Height: winsize.Height}
+}
+
+func SetSize(fd uintptr, size Size) error {
+	return term.SetWinsize(fd, &term.Winsize{Height: size.Height, Width: size.Width})
+}
+
+func (t *TTY) MonitorSize(initialSizes ...*Size) {
+	// make it buffered so we can send the initial terminal sizes without blocking, prior to starting
+	// the streaming below
+	t.resizeChan = make(chan Size, len(initialSizes))
+	t.stopResizing = make(chan struct{})
+
+	for i := range initialSizes {
+		if initialSizes[i] != nil {
+			t.resizeChan <- *initialSizes[i]
+		}
+	}
+
+	resizeSignals := make(chan os.Signal, 1)
+	signal.Notify(resizeSignals, syscall.SIGWINCH)
+
+	go func() {
+		defer runtime.HandleCrash()
+
+		for {
+			select {
+			case <-resizeSignals:
+				size := t.GetSize()
+				if size == nil {
+					continue
+				}
+
+				select {
+				case t.resizeChan <- *size:
+				default:
+				}
+			case <-t.stopResizing:
+				return
+			}
+		}
+	}()
+}
+
+type TerminalSizeQueue interface {
+	Next() *Size
+}
+
+func (t *TTY) Next() *Size {
```

Comment:
yeah.

On Fri, Apr 22, 2016 at 3:51 PM, Andy Goldstein <notifications@github.com>
wrote:

> In pkg/util/term/term.go
> <https://github.com/ncdc/kubernetes/pull/7#discussion_r60793866>:
>
> > +				select {
> > +				case t.resizeChan <- *size:
> > +				default:
> > +				}
> > +			case <-t.stopResizing:
> > +				return
> > +			}
> > +		}
> > +	}()
> > +}
> > +
> > +type TerminalSizeQueue interface {
> > +	Next() *Size
> > +}
> > +
> > +func (t *TTY) Next() *Size {
>
> @smarterclayton <https://github.com/smarterclayton> I'm not super crazy
> about TTY having a method on it called Next. Maybe I should create type
> defaultTerminalSizeQueue struct and have TTY delegate to that instead.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/ncdc/kubernetes/pull/7/files/4da1819c601623c127cf1e30d6544565a8c82419#r60793866>
>